vulnerabilities/caching-storage-in-memory.md

```diff
@@ -0,0 +1,126 @@
+## Caching Storage in Memory
+
+Storage variables such as structs are often loaded into memory for efficiency. When incorrectly handled, this can lead to inconsistent state updates. This type of vulnerability can manifest when developers cache storage values in memory variables but either fail to update the storage after modifying the memory variables or incorrectly synchronize between memory and storage.
+
+### Storage vs Memory in Solidity
+
+In Solidity, `storage` refers to data that is persistently stored on the blockchain, while `memory` is temporary and exists only during function execution. Reading from storage is expensive, so developers often cache storage values in memory to reduce gas costs. However, this optimization can lead to vulnerabilities if not handled correctly. 
+
+### Self-transfer Free Minting Bug
+
+One common example of this vulnerability is the "self-transfer free minting" bug, where tokens can be minted for free by exploiting incorrect memory handling during self-transfers.
+
+```solidity
+struct UserData {
+    bytes32 id;
+    uint256 balance;
+    address account;
+}
+
+mapping(address => UserData) public users;
+
+// VULNERABLE
+function transfer(address from, address to, uint248 transferValue)  {   
+    UserData memory _accountFrom = users[from];
+    if (blacklisted[_accountFrom.id]) revert();
+    
+    UserData memory _accountTo = users[to];
+    if (blacklisted[_accountTo.id]) revert();
```

Comment:
nit: We should drop these blacklisted checks since they're not relevant to the vulnerability. Could also then drop `id` from `UserData`